File: 21. Merge Two Sorted Lists.cpp
#include <iostream>
#include
 //Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    // O(M+N), recursive: 22.89%, ok
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        if(l1 == NULL) return l2;   // boundary conditions
        if(l2 == NULL) return l1;
        if(l1 -> val <= l2 -> val){
            l1 -> next = mergeTwoLists(l1 -> next, l2);
            return l1;
        }
        else{
            l2 -> next = mergeTwoLists(l1, l2 -> next);
            return l2;
        }
    }

    // O(M+N), iterative: 22.89%, ok
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        if(!l1) return l2;
        if(!l2) return l1;
        ListNode ret(0);       // bug: ListNode* ret = l
        ListNode* l = &ret;    // bug: ListNode* l
        while(l1 && l2){
            if(l1->val <= l2->val) {
                l->next = l1;
                l1 = l1->next;
            }
            else {
                l->next = l2;
                l2 = l2->next;
            }
            l = l->next;
        }
        if(l1) l->next = l1;
        if(l2) l->next = l2;
        return ret.next;
    }
};

File: 21. Merge Two Sorted Lists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    
    # iterative: O(m+n) time &  space
    def mergeTwoLists(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        ret = ListNode(-1)
        dummy = ret
        while l1 and l2:
            if l1.val <= l2.val:
                dummy.next = ListNode(l1.val)
                l1, dummy = l1.next, dummy.next
            else:
                dummy.next = ListNode(l2.val)
                l2, dummy = l2.next, dummy.next
        while l1:
            dummy.next = ListNode(l1.val)
            l1, dummy = l1.next, dummy.next
        while l2:
            dummy.next = ListNode(l2.val)
            l2, dummy = l2.next, dummy.next
        return ret.next
    
    # more elegant iterative: O(m+n) time & space
    def mergeTwoLists(self, l1, l2):
        ret = ListNode(-1)
        dummy = ret
        while l1 and l2:
            if l1.val <= l2.val:
                dummy.next, l1 = ListNode(l1.val), l1.next
            else:
                dummy.next, l2 = ListNode(l2.val), l2.next
            dummy = dummy.next
            
        dummy.next = l1 if l1 else l2
        return ret.next

    # recursive: O(m+n) time & space
    def mergeTwoLists(self, l1, l2):
        if not l1 or not l2:
            return l1 or l2
        if l1.val <= l2.val:
            l1.next = self.mergeTwoLists(l1.next, l2)
            return l1
        else:
            l2.next = self.mergeTwoLists(l1, l2.next)
            return l2